application/controllers/handler_lti.js

```diff
@@ -3,118 +3,147 @@
 const boom = require('boom'), //Boom gives us some predefined http codes and proper responses
   co = require('../common'),
   userCtrl = require('../database/user'),
-  //ltiCtrl = require('../database/lti'),
+  userltiCtrl = require('../database/userlti'),
   //config = require('../configuration'),
   jwt = require('./jwt'),
   util = require('./util'),
   lti = require('ims-lti'),
-  LTI_CONFIG = require('../configuration.js').LTI;
+  LTI_ID = require('../configuration.js').LTI_ID;
 
 const PLATFORM_LTI_URL = require('../configs/microservices').platform.uri+ '/ltiLogin';
 
 module.exports = {
 
   handleLTI: (req, res) => {
-    //console.log('req='+JSON.stringify(req.query));
-
-    if (!LTI_CONFIG.ENABLED)
-      return res(boom.notFound());
-
-    // Validate LTI request
-    let ltiKeySecret = {
-      '_id': LTI_CONFIG._ID,
-      'key': LTI_CONFIG.KEY,
-      'secret': LTI_CONFIG.SECRET
-    };
-
-    //TODO: Accept different types of signatures
-    let ltiProvider = new lti.Provider(ltiKeySecret.key, ltiKeySecret.secret);
-    ltiProvider.valid_request(req, function(err, isValid){
-      if(err){
-        console.log('There was an error in the LTI request', err);
-        res(boom.badImplementation());
-      }
-      else{
-        let user = getUser(req);
 
-        console.log('Registration with LTI: ', user.username, user.email);
-        //check if username already exists
-        return util.isIdentityAssigned(user.email, user.username)
-          .then((result) => {
+    let resource_id = req.params.resource_id;
+    let isValid = false;
+
+    return userltiCtrl.readAllLTIs()
+      .then((ltiArray) => {
+        //console.log("ltiArray.length="+ltiArray.length);
+        let found = false;
+        for(let i=0; i<ltiArray.length; i++){
+          if(!found){
```

Comment:
There are asynchornous function calls within the synchronous loop, and the flag is there to break the loop once the the relevant asychronous function call is made.

Reply:
I see what was you intention but this flag has no effect because the loop starts all requests (ltiProvider.valid_request) which are then executed concurrent (because of the callback). Thus for each element of the array the code gets executed. I dont know how expensive ltiProvider.valid_request is but executing it for every entry could be a heavy task.
At the end you have to decide if you just call it for every element in the loop (much requests from the user-service at a time) or do everything sequencial(request against the user-service could take long) depending on what you think is the best evil. Or you use something out of asyncJS or so to stop all remaining processes when you have found the correct one.